connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/SparkSession.scala

```diff
@@ -813,6 +823,23 @@ class SparkSession private[sql] (
    * Set to false to prevent client.releaseSession on close() (testing only)
    */
   private[sql] var releaseSessionOnClose = true
+
+  private[sql] def registerObservation(planId: Long, observation: Observation): Unit = {
+    if (observationRegistry.containsKey(planId)) {
+      throw new IllegalArgumentException("An Observation can be used with a Dataset only once")
+    }
+    observationRegistry.put(planId, observation)
+  }
+
+  private[sql] def setMetricsAndUnregisterObservation(
+      planId: Long,
+      metrics: Map[String, Any]): Unit = {
+    if (observationRegistry.containsKey(planId)) {
```

Comment:
From a threading point of view i'd be better not to touch the map twice. Just try to remove the observation for the planId, and check if it is not null.